File: backend/app/models/maze.py
from pydantic import BaseModel, Field
from typing import List, Optional
from enum import Enum


class MazeType(str, Enum):
    PERFECT = "perfect"
    LOOP = "loop"
    BRAID = "braid"


class MazeAlgorithm(str, Enum):
    BACKTRACKING = "backtracking"
    PRIM = "prim"
    KRUSKAL = "kruskal"
    ELLER = "eller"
    WILSON = "wilson"


class Cell(BaseModel):
    row: int
    col: int


class MazeGenerationRequest(BaseModel):
    rows: int = Field(..., gt=4, description="Number of rows in the maze (min 5)")
    cols: int = Field(..., gt=4, description="Number of columns in the maze (min 5)")
    algorithm: MazeAlgorithm
    maze_type: MazeType = MazeType.PERFECT


class MazeResponse(BaseModel):
    maze: List[List[int]]
    steps: List[List[List[int]]]  # Animation steps

File: frontend/src/types/maze.ts
// Cell represents a position in the maze
export interface Cell {
  row: number
  col: number
}

// Maze types
export enum MazeType {
  PERFECT = 'perfect',
  LOOP = 'loop',
  BRAID = 'braid',
}

// Maze generation algorithms
export enum MazeAlgorithm {
  BACKTRACKING = 'backtracking',
  PRIM = 'prim',
  KRUSKAL = 'kruskal',
  ELLER = 'eller',
  WILSON = 'wilson',
}

// Maze data structure
export type Maze = number[][]

// Animation steps for maze generation
export type MazeSteps = number[][][]

// Response from maze generation API
export interface MazeResponse {
  maze: Maze
  steps: MazeSteps
}

// Request to generate a maze
export interface MazeGenerationRequest {
  rows: number
  cols: number
  algorithm: MazeAlgorithm
  maze_type: MazeType
}
